pyat/at/physics/dynamic_aperture.py

```diff
@@ -0,0 +1,1012 @@
+                              )
+                    # make sure 0.0 is in the scan
+                    # if 0.0 not in d_[-1]:
+                    #    d_[-1].append(0.0);
+                else:  # this dimension is not part of the scan
+                    d_.append((self.dict_def_range[p][0] +
+                               self.dict_def_range[p][1])/2
+                              )
+                    # make sure 0.0 is in the scan
+                    # if d_ is not 0.0:
+                    #    d_.append(0.0);
+
+            # define mesh of points
+            xx, xpxp, yy, ypyp, deltadelta, ctct = np.meshgrid(d_[0],
+                                                               d_[1],
+                                                               d_[2],
+                                                               d_[3],
+                                                               d_[4],
+                                                               d_[5],)
+
+            self.coordinates['x'] = xx.flatten()
+            self.coordinates['xp'] = xpxp.flatten()
+            self.coordinates['y'] = yy.flatten()
+            self.coordinates['yp'] = ypyp.flatten()
+            self.coordinates['delta'] = deltadelta.flatten()
+            self.coordinates['ct'] = ctct.flatten()
+
+        if len(self.coordinates['x']) == 0:
+            raise IOError('grid mode must be grid or radial')
+
+        self.survived = [False]*len(self.coordinates['x'])
+
+        pass
+
+    def test_survived(self, coordinates):
+        """
+        test if a set of particle coordinates survived
+        returns a boolean if the coordinates survived
+        """
+
+        """
+        if delta != 0.0:
+            # recompute orbit
+            self.ring.radiation_off()
+            orbit_delta = self.ring.find_sync_orbit()
+            self.ring.radiation_on()
+
+            test_coord = list(orbit_delta[0])
+        else:
+            test_coord = list(self.orbit[0])
+        """
+
+
+        # nothing to do if no coordinate to test
+        if len(coordinates) == 0:
+            return []
+
+        # if coordinates to test:
+
+        # create 6xN matrix add orbit (with dpp) to each coordinate
+        rin = np.concatenate(([coordinates['x'] + self.orbit[0][0]],
+                              [coordinates['xp'] + self.orbit[0][1]],
+                              [coordinates['y'] + self.orbit[0][2]],
+                              [coordinates['yp'] + self.orbit[0][3]],
+                              [coordinates['delta'] + self.orbit[0][4]],
+                              [coordinates['ct'] + self.orbit[0][5]]),
+                              axis=0)
+
+        # track coordinates for N turns
+        if not self.parallel_computation:
+
+            t, losses = at.atpass(self.ring,
+                                  copy.deepcopy(np.asfortranarray(rin)),
+                                  self.number_of_turns,
```

Comment:
A deepcopy of a numerical array does not make sense !
```python
            t, losses = at.atpass(self.ring,
                                  rin.copy(),
                                  self.number_of_turns,

```

Reply:
here I needed to add np.asfrotranarray. it is not needed?

Reply:
in these calls, I also have the flag losses = True. Is there a way to minimize the output in this flag? I remember that in the matlab version it is a rather big output, and here I just need an array of booleans.

Reply:
> here I needed to add np.asfrotranarray. it is not needed?

  File "/Users/liuzzo/beamdyn/at/pyat/at/physics/dynamic_aperture.py", line 529, in test_survived
    t, losses = at.atpass(self.ring,
ValueError: rin is not Fortran-aligned

Reply:
The losses flag was meant to compute loss maps. 
In case you just need a vector of boolean I would suggest the following   

`mask = numpy.isnan(t[0,:,0,-1])`

Or something similar: a lots particle will have its first coordinate set to nan

Reply:
Dear @swhite2401 ,

thanks for your reply. Infact this is what I used to do, before reading that atpass has losses output. So, I thought it would have been a good choice at least for readibility, to use losses. 
Actually, in the matlab ringpass there are 3 outputs, [ particlescoordinates, loss_flag, losses ]= ringpass(...). 
In pyAT there is a ringpass equivalent? 

If it is faster/better/suggested I will do as you say:

`mask = numpy.isnan(t[0,:,0,-1])
`

where t[coordinate, particle, turn, index_in_lattice ]

I guess the t shape/dimensions will be the same when tracking a single particle or multiple particles

Reply:
Yes `lattice_pass` also has the flag available, I ported everything expect  for the history buffer of lost particles (that was the big chunk). In fact activating the flag represent a negligible loss in performance.
I am not quite sure about your coordinate order... it seems the last one is the turn number which is why I did put -1 there.
No difference for multi-particle.

Reply:
ok. so t[coordinate, particle, index_in_lattice, turn_number]
I commit in few minutes the change

Reply:
Here is the help, source is in tracking/patpass.py     

OUTPUT:
        (6, N, R, T) array containing output coordinates of N particles
        at R reference points for T turns.
        If losses ==True: {islost,turn,elem,coord} dictionnary containing
        flag for particles lost (True -> particle lost), turn, element and
        coordinates at which the particle is lost. Set to zero for particles
        that survived

Reply:
By the way if you have doubts about a function you can easily access the help in a python prompt by typing:
`help(at.tracking.patpass)` for example

Reply:
finally I have committed this:

notice the np.asfortranarray()

```

t = at.atpass(self.ring,
                                  np.asfortranarray(rin.copy()),
                                  self.number_of_turns,
                                  refpts=np.array(np.uint32(0)))

t = at.patpass(self.ring,
                       rin.copy(),
                       self.number_of_turns,
                       refpts=np.array(np.uint32(0)))

```

Reply:
I think you can use the default `refpts=None` in case you want the coordinates at the start of the ring only.

For single core I would suggest to use `lattice_pass` instead (no need to care about ordering in this case), I will add the same check in patpass so that you don't need to worry about this anymore

Reply:
using ..., refpts=None) I get this error:

TypeError: argument 4 must be numpy.ndarray, not None

Reply:
removing refpts, it gives an errors when looking at the output tracking, the refpts dimension is empty

survived = [not s for s in np.isnan(t[0,:,0,-1])]
IndexError: index 0 is out of bounds for axis 2 with size 0

Reply:
so for now I keep it like it is

Reply:
Ah yes this is in fact an problem I have noticed in the past, using the default (removing refpts),
It probably needs fixing

Reply:
if you noticed and I notice, may be we should open an issue?

Reply:
**You must not use `atpass` directly**:
the public interface is `lattice_pass`. The arguments are the same, but `lattice_pass` does all the argument checking that was too tricky to do in C. This answers both you questions concerning alignment and refpts.

Reply:
Coming back to Fortran alignment: It is not strictly necessary, since both lattice_pass and patpass will reorganise the memory if necessary. However it's more efficient if you set Fortran alignment in r_in, so that the memory layout is done once for ever. This is true for both `lattice_pass` and `patpass`.

In your particular case, this can be done easily by:
```
        # create 6xN matrix add orbit (with dpp) to each coordinate
        rin = np.asfortranarray(list(coordinates.values())) + self.orbit.reshape((6,1))
```

Reply:
Dear @lfarv 

I replaced this: 

```
 rin = np.concatenate(([coordinates['x'] + self.orbit[0][0]],
                              [coordinates['xp'] + self.orbit[0][1]],
                              [coordinates['y'] + self.orbit[0][2]],
                              [coordinates['yp'] + self.orbit[0][3]],
                              [coordinates['delta'] + self.orbit[0][4]],
                              [coordinates['ct'] + self.orbit[0][5]]),
                              axis=0)

```

with 

```

rin = np.array(list(coordinates.values()))

 p = list(rin.shape)
 p[0] = 1
 orb =  np.tile(self.orbit[0],p)

```

in order to make it work also in 2D.
Both options would be eventually followed by :

`rin = np.asfortranarray(rin)
`

However, I have the impression that the first version is actually more intuitive to read.

Reply:
Are you sure this is the case for `patpass`? I am not convinced this is the case...to checked

Reply:
Dear @lfarv 

I replaced this: 

```
 rin = np.concatenate(([coordinates['x'] + self.orbit[0][0]],
                              [coordinates['xp'] + self.orbit[0][1]],
                              [coordinates['y'] + self.orbit[0][2]],
                              [coordinates['yp'] + self.orbit[0][3]],
                              [coordinates['delta'] + self.orbit[0][4]],
                              [coordinates['ct'] + self.orbit[0][5]]),
                              axis=0)

rin = np.asfortranarray(rin)

```

with 

```
rin = np.array(list(coordinates.values()))

p = list(rin.shape)
p[0] = 1
orb =  np.tile(np.array([self.orbit[0]]).T,p)

rin = np.asfortranarray(rin+orb)

```

in order to make it work also in 2D.

However, I have the impression that the first version is actually more intuitive to read.

Reply:
> Are you sure this is the case for `patpass`? I am not convinced this is the case...to checked

Yes, sure:
```python
args = [(ring, r_in[:, i]) for i in range(r_in.shape[1])]
```
` r_in[:, i]` consecutive implies Fortran ordering

Reply:
Ok, but still it fails in case a collective `passmethod` is present or for one particle.
We may just force it in a similar fashion to what is done in `lattice_pass` to avoid problems

By the way, would you know any smart way to prevent people from using atpass directly? I have seen that AT users tend to do it a lot...

Reply:
> ```
> rin = np.array(list(coordinates.values()))
> 
> p = list(rin.shape)
> p[0] = 1
> orb =  np.tile(np.array([self.orbit[0]]).T,p)
> 
> rin = np.asfortranarray(rin+orb)
> ```
Can you explain what you are trying to do? `self.orbit `is supposed to have shape (6,): that is a 1D array. I couldn't find where a 2nd dimension appears.

And what is wrong with my version?

np.array(list(coordinates.values())) has shape (6,n)
self.orbit.reshape((6,1)) has shape (6,1)
So where is the problem?

Reply:
> Ok, but still it fails in case a collective `passmethod` is present or for one particle.
> We may just force it in a similar fashion to what is done in `lattice_pass` to avoid problems

Well, I don't think it's worth changing anything. This memory rearrangement is minor compared to the overhead of starting the thread… And forcing alignment earlier will not make it faster.

> By the way, would you know any smart way to prevent people from using atpass directly? I have seen that AT users tend to do it a lot...

May be:
1. renaming it to `_atpass`, showing clearly it's private
2. Instead of importing it in `__init__.py`, importing it where it's needed (`track.py` etc.)
3. Adding a dummy `atpass` raising an exception and telling what to do

Remember, I suggested `patpass` could be renamed as `plattice_pass` to better show the equivalence.

Reply:
This is probably a good time to do this what do you think? 
We can leave the original methods with a deprecation warning until the next version increment

Reply:
> > ```
> > rin = np.array(list(coordinates.values()))
> > 
> > p = list(rin.shape)
> > p[0] = 1
> > orb =  np.tile(np.array([self.orbit[0]]).T,p)
> > 
> > rin = np.asfortranarray(rin+orb)
> > ```
> 
> Can you explain what you are trying to do? `self.orbit `is supposed to have shape (6,): that is a 1D array. I couldn't find where a 2nd dimension appears.
> 
> And what is wrong with my version?
> 
> np.array(list(coordinates.values())) has shape (6,n)
> self.orbit.reshape((6,1)) has shape (6,1)
> So where is the problem?

it says it cannot "broadcast" , so I tried several .T, tile, shape, reshape, till I got to the one I sent.

Reply:
> Ok, but still it fails in case a collective `passmethod` is present or for one particle.
> We may just force it in a similar fashion to what is done in `lattice_pass` to avoid problems
> 
> By the way, would you know any smart way to prevent people from using atpass directly? I have seen that AT users tend to do it a lot...

I did this! nothing forbid me, or warned me I was doing something not welcome. An _atpass would have discouraged me.

In case, this modification would be a different PR, right?

Reply:
> > > ```
> > > rin = np.array(list(coordinates.values()))
> > > 
> > > p = list(rin.shape)
> > > p[0] = 1
> > > orb =  np.tile(np.array([self.orbit[0]]).T,p)
> > > 
> > > rin = np.asfortranarray(rin+orb)
> > > ```
> > 
> > 
> > Can you explain what you are trying to do? `self.orbit `is supposed to have shape (6,): that is a 1D array. I couldn't find where a 2nd dimension appears.
> > And what is wrong with my version?
> > np.array(list(coordinates.values())) has shape (6,n)
> > self.orbit.reshape((6,1)) has shape (6,1)
> > So where is the problem?
> 
> 
> it says it cannot "broadcast" , so I tried several .T, tile, shape, reshape, till I got to the one I sent.

Can you look where the shapes differ from what I expect and mentioned above?

Reply:
rin.shape
(6, 169)
self.orbit[0].shape
(6,)
rin + self.orbit[0]
Traceback (most recent call last):
  File "/Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
ValueError: operands could not be broadcast together with shapes (6,169) (6,)

Reply:
So you have to use `rin + self.orbit[0].reshape((6,1))`. Broadcasting prepends the missing dimensions, so reshapes orbit[0] as (1,6) instead of (6,1), and addition fails.

But I still don't see the shape of `self.orbit` itself, and why you use `self.orbit[0]` instead of `self.orbit`.

Reply:
I use self.orbit[0] due to this:
```

>> self.orbit
(array([-2.63484764e-09, -1.45845243e-10,  0.00000000e+00,  0.00000000e+00,
       -6.69657366e-06, -5.88436654e-02]), array([], shape=(0, 6), dtype=float64))
>> self.orbit[0]
array([-2.63484764e-09, -1.45845243e-10,  0.00000000e+00,  0.00000000e+00,
       -6.69657366e-06, -5.88436654e-02])
>> self.orbit[0]+rin   # this works
array([ 9.94730305e-07, -2.91690487e-10,  1.00000000e-06,  0.00000000e+00,
       -1.33931473e-05, -1.17687331e-01])
>> self.orbit+rin   # this gives error
<input>:1: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
Traceback (most recent call last):
  File "/Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
ValueError: operands could not be broadcast together with shapes (2,) (6,) 


```

Reply:
This seems to work:

```
rin = np.array(list(coordinates.values()))

if len(rin.shape) == 1:
    rin = rin + self.orbit[0]
else:
    rin = rin + self.orbit[0].reshape(6,1)

```

Reply:
Now, I understand everything ! When calling `find_orbit`, you get 2 output arguments: `orbit0` and `orbit`. So the call should be:

`self.orbit, _ = find_orbit(...)`

As it is now, `self.orbit` is a tuple with 2 values. `orbit0` and an empty array. And I missed that !

Reply:
The final line should be:

`rin = np.asfortranarray(list(coordinates.values()).reshape(6, -1) + self.orbit.reshape(6, 1)`

By the way, `len(rin.shape)` is `rin.ndim`

Reply:
that's python! I do the modification